examples_tests/41.VisibilityBuffer/main.cpp

```diff
@@ -3,815 +3,402 @@
 // For conditions of distribution and use, see copyright notice in nabla.h
 
 #define _NBL_STATIC_LIB_
-#include <nabla.h>
-
-#include "../common/QToQuitEventReceiver.h"
-#include <nbl/video/IGPUVirtualTexture.h>
-#include <nbl/asset/CMTLPipelineMetadata.h>
-#include <nbl/asset/filters/CMipMapGenerationImageFilter.h>
-
 #include <iostream>
 #include <cstdio>
+#include <nabla.h>
 
+//! I advise to check out this file, its a basic input handler
+#include "../common/QToQuitEventReceiver.h"
+#include "nbl/asset/utils/CCPUMeshPackerV1.h"
+#include "nbl/asset/CCPUMeshPackerV2.h"
 
 using namespace nbl;
 using namespace core;
+using namespace asset;
+using namespace video;
 
-
-constexpr const char* GLSL_COMMON_OVERRIDE =
+constexpr const char* SHADER_OVERRIDES[2] =
+{
 R"(
-#extension GL_NV_fragment_shader_barycentric : enable
-#extension GL_AMD_shader_explicit_vertex_parameter : enable
+#define _NBL_VERT_INPUTS_DEFINED_
+#define _NBL_VERT_MAIN_DEFINED_
+#define _NBL_VERT_OUTPUTS_DEFINED_
+#define _NBL_VERT_SET1_BINDINGS_DEFINED_
 
-#if !defined(IRR_GL_NV_fragment_shader_barycentric) && !defined(GL_AMD_shader_explicit_vertex_parameter) 
-    #error "GL_NV_fragment_shader_barycentric, nor GL_AMD_shader_explicit_vertex_parameter available on your GPU, or IrrlichtBaW messed up the support"
-#endif
+#define VirtualAttribute_t uvec2
 
-#ifdef GL_AMD_shader_explicit_vertex_parameter
-    #error "Don't have AMD hardware to test https://gpuopen.com/stable-barycentric-coordinates/"
-#endif
+vec4 nbl_glsl_decodeRGB10A2_UNORM(in uint x)
+{
+	const uvec3 rgbMask = uvec3(0x3ffu);
+	const uvec4 shifted = uvec4(x,uvec3(x)>>uvec3(10,20,30));
+	return vec4(vec3(shifted.rgb&rgbMask),shifted.a)/vec4(vec3(rgbMask),3.0);
+}
 
-struct PerInstanceData
+vec4 nbl_glsl_decodeRGB10A2_SNORM(in uint x)
 {
-    mat4 modelViewProj;
-    vec3 Ka;
-    vec3 Kd;
-    vec3 Ks;
-    vec3 Ke;
-    uvec2 map_Ka_data;
-    uvec2 map_Kd_data;
-    uvec2 map_Ks_data;
-    uvec2 map_Ns_data;
-    uvec2 map_d_data;
-    uvec2 map_bump_data;
-    float Ns;
-    float d;
-    float Ni;
-    uint extra; //flags copied from MTL metadata
-};
+    uvec4 shifted = uvec4(x,uvec3(x)>>uvec3(10,20,30));
+    const uvec3 rgbMask = uvec3(0x3ffu);
+    const uvec3 rgbBias = uvec3(0x200u);
+    return max(vec4(vec3(shifted.rgb&rgbMask)-rgbBias,float(shifted.a)-2.0)/vec4(vec3(rgbBias-uvec3(1u)),1.0),vec4(-1.0));
```

Comment:
not sure, but having doubts about this
If it still doesnt work, consider this
```glsl
ivec3 rgb = ivec3(x)>>ivec3(0,10,20);
rgb = bitfieldExtract(rgb, 0, 10);
vec3 rgb_ = vec3(rgb)/511.0;
float a_ = bitfieldExtract(int(x),30,2);
return vec4(rgb_,a_);
```

Reply:
yeah the bitshifts in the original are definitely messed up, use @Crisspl's version but its also wrong

it should have a -1.0 after the division in all components and a `max(,vec4(-1.0))`